File: Astro-TFG/public/scripts/1-calculos-matematicos/prueba-1/version-2/python/main.py
import gc
import time
import js  # type: ignore
import tracemalloc
import numpy as np
from pyscript import display

seed = int(time.time() * 1000) % 2**32
rng = np.random.default_rng(seed)


def create_matrix(size):
    return rng.random((size, size))


def get_memory_usage():
    return tracemalloc.get_traced_memory()[1] / (1024 * 1024)


def multiply_matrices(size):
    tracemalloc.start()

    A = create_matrix(size)
    B = create_matrix(size)

    start = time.perf_counter()
    gc.collect()

    C = np.dot(A, B)

    end = time.perf_counter()
    execution_time = (end - start) * 1000

    memory_usage = get_memory_usage()
    tracemalloc.stop()

    results = {
        "execution_time": execution_time,
        "memory_usage": memory_usage
    }

    display_results(results, size)


def display_results(results, size):
    output_element = js.document.getElementById("pyscript-output")
    output_element.innerHTML += f"""
        <div style="font-weight: bold;">Matriz {size}x{size}</div>
        """
    display(f"ET: {results['execution_time']:.2f} ms",
            target="pyscript-output")
    display(f"RAM: {results['memory_usage']:.2f} MB", target="pyscript-output")
    js.endTimerWebAssembly()
    output_element.appendChild(js.document.createElement('br'))
    output_element.appendChild(js.document.createElement('hr'))
    output_element.appendChild(js.document.createElement('br'))


def run_py_benchmark(event):
    js.clearCell("pyscript-output")

    multiply_matrices(500)
    multiply_matrices(1000)
    multiply_matrices(2000)


js.window.run_py_benchmark = run_py_benchmark

File: Astro-TFG/public/scripts/1-calculos-matematicos/prueba-1/version-2/javascript/main.js
async function multiplyMatrices(size) {
    let A = tf.randomUniform([size, size], 0.0, 1.0);
    let B = tf.randomUniform([size, size], 0.0, 1.0);

    const start = performance.now();

    let C = tf.matMul(A, B);

    const executionTime = performance.now() - start;

    const results = {
        size: `${size}x${size}`,
        execution_time: executionTime,
        memory_usage: getMemoryUsageJS(),
    };

    displayResults(results);
}

function getMemoryUsageJS() {
    if (performance.memory) {
        return Math.max(performance.memory.usedJSHeapSize / (1024 * 1024), 0);
    }
    return -1;
}


function displayResults(results) {
    const output = document.getElementById("javascript-output");

    const titleDiv = document.createElement("div");
    titleDiv.textContent = `Matriz ${results.size}`;
    titleDiv.style.fontWeight = "bold";
    output.appendChild(titleDiv);

    const timeDiv = document.createElement("div");
    timeDiv.textContent = `ET: ${results.execution_time.toFixed(2)} ms`;
    output.appendChild(timeDiv);

    const memoryDiv = document.createElement("div");
    if (results.memory_usage !== -1) {
        memoryDiv.textContent = `RAM: ${results.memory_usage.toFixed(2)} MB`;
    } else {
        memoryDiv.textContent = `RAM measurement unsupported in this browser.`;
    }
    output.appendChild(memoryDiv);
    output.appendChild(document.createElement("br"));
    output.appendChild(document.createElement("hr"));
    output.appendChild(document.createElement("br"));
}

window.runJsBenchmark = async function () {
    window.clearCell("javascript-output");
    window.showExecutionLoader();

    requestAnimationFrame(() => {
        setTimeout(async () => {
            await multiplyMatrices(500);
            await multiplyMatrices(1000);
            await multiplyMatrices(2000);

            window.hideExecutionLoader();
        }, 0);
    });
};